### DISCUSSION TITLE
Slow speed iscsi or aoe

### DISCUSSION THREAD

**Discussion by boris-dibrov** [ipxe/ipxe#241] [Q&A]
### Hello!

_(Sorry for my english)_

We use ipxe to boot client machines over the network.  

---  

### Our working conditions:

1. DHCP and TPTP on a host running Windows 10
2. Iscsi server Starwind on the same computer
3. Iscsi target. (Disk format - NTFS. Loader – Grub4Dos. Windows 7 – vhd)
4. Gigabit network

### Scheme of work:
Starting the computer -> loading the ipxe over the network -> connecting to the iscsi –> grub4dos load the vhd image (Windows 7) in RAMDisk and chainloader to windows bootmgr

### Problem:
When the Grub4Dos load the vhd image in ramdisk on many computers everything works ok. The link speed is 650 megabits. But on other computers the link speed is max 100 megabits (1gb port). And it constantly jumps from 40 to 100. The experiments were carried out on the same network and with the same host. One client loads at a speed of 650 another 40-50.

### What have we tried:
1. Compile the latest version of ipxe. (.pxe, .kpxe, .kkpxe) for i386-pcibios and for x86_x64-pcibios
2. We tried ipxe.iso from usb. Exclude dhcp and tftp
3.  We tried to in efi mode
4. We tried update bios

The result is the same. We need high speed, our vhd image is 5 gigabytes. And at high speed, the computer starts up in an average of 1 minute.

### Examples of equipment (All standard, latest BIOS versions):

_Everything OK (650 mb/s):_

1. Motherboard - Gigabyte F2-A68HM-S1
2. Motherboard – ASUS H110M-R
3.  Motherboard – ASUS A68HM-K
4. Motherboard – MSI H81M-P33
5.  Motherboard – AsRock Z370M Pro4
6.  Motherboard –Gigabyte H310M S2

_Not OK (40 – 100 mb/s):_

1. Motherboard – ASUS Prime A320M-R
2. Motherboard – Gigabyte ga-a320m-s2h
3. Motherboard – ASUS J1800I-C
4. Motherboard – GIGABYTE J1800N-D2H
5. Motherboard – AsRock D1800B-ITX

----------------------------------------------------

All computers have gigabit ports. All experiments were carried out on the same network and with one host.

Also we tried Ubuntu. If the host is ubuntu, then the iscsi the speed is 10 megabits, and the aoe is 100 megabits. Although the network is gigabit and if you connect to the iscsi target from Windows or Ubuntu, then the speed is 1000 megabits.

**Comment by NiKiZe** [comment]
run `ifstat` in iPXE
especially collect data on which NICs these machines have. - also collect the actual link speed, get the data from the switch
Compare both working and non working

You could also try to do downloads and then run ifstat again, and check if there is any increase in RXE TXE values

**Reply by boris-dibrov** [reply]
Now i have only two motherboards. Next week, I can do checks on everyone.
All network interfaces are integrated into the motherboard. Need to see the network interface model on the motherboard?

### ifstat

1. Gigabyte F2-A68HM-S1 - **works OK**
![photo_2021-01-31_18-48-42](https://user-images.githubusercontent.com/46572522/106389566-fc3da300-63f4-11eb-9bae-6de7cb39c3b3.jpg)

---

2. Gigabyte ga-a320m-s2h - **not OK**
![photo_2021-01-31_18-49-38](https://user-images.githubusercontent.com/46572522/106389582-15deea80-63f5-11eb-8baa-b6f8d82d2bfb.jpg)

**Reply by mcb30** [reply]
It looks potentially similar to #145 

@BorisDibrov How reproducible is this?  For one of the "slow" machines, do you always get a slow speed?

**Reply by boris-dibrov** [reply]
If I connect to the Starwind target iscsi, then some machines operate at 650 megabits, some from 40 to 100 megabits. If I connect to the target on ubuntu (tgt), then the speed on all machines is 10 megabits. If you connect to ubuntu via AoE, then the speed is 100 megabits on all machines

**Reply by boris-dibrov** [reply]
> It looks potentially similar to #145
> 
> @BorisDibrov How reproducible is this? For one of the "slow" machines, do you always get a slow speed?

Yes, it happens all the time. I tried to exclude the network equipment and connect two machines directly, the result is the same

**Reply by NiKiZe** [reply]
Common here is rtl8168 as mcb30 points out.

Also please try to be clear if you are talking about actual ethernet link speed, or if you are talking about observed transfer speed.

**Reply by boris-dibrov** [reply]
> Common here is rtl8168 as mcb30 points out.
> 
> Also please try to be clear if you are talking about actual ethernet link speed, or if you are talking about observed transfer speed.

Transfer spped

**Reply by boris-dibrov** [reply]
I thought that the bottleneck might be the processor, since the hardware is weak. But no, grub4dos loads from a 3.0 flash drive into RAM in about 1 minute the vhd (5 gigabyte) on all machines

**Reply by boris-dibrov** [reply]
I can do all the tests, just explain which ones are needed and how to do them. I'm new to booting over the net, have never done this, so sorry for the stupid questions. I can also provide access to equipment, if needed

**Reply by NiKiZe** [reply]
As I wrote, please grab the actual link speed and duplex, your switch should be able to tell you.

**Reply by boris-dibrov** [reply]
The actual link speed is 1 gigabit. On windows ethernet adapter shows 1 gigabit. On rj-45 lan ports **Orange** LED. Orange color indicates that the network is operating in 1 gigabit mode (Taken from the documentation from the manufacturer's website). Or are you not talking about that?

**Reply by NiKiZe** [reply]
Link speed when running iPXE? All of them run at gig speed both working and "not working" devices?

**Reply by boris-dibrov** [reply]
> Link speed when running iPXE? All of them run at gig speed both working and "not working" devices?

Yes, on all gig. And nic on all machines rtl8168

**Reply by boris-dibrov** [reply]
The problem is similar to this https://forum.ipxe.org/showthread.php?tid=7807. Since if I run Windows directly from iscsi, the Windows logo hangs for about 1 minute. I tried the suggested solutions in this post, didn't help

**Reply by mcb30** [reply]
> The problem is similar to this https://forum.ipxe.org/showthread.php?tid=7807. Since if I run Windows directly from iscsi, the Windows logo hangs for about 1 minute. I tried the suggested solutions in this post, didn't help

@BorisDibrov Can you be specific about which suggested solutions you have tried?

**Reply by boris-dibrov** [reply]
> > The problem is similar to this https://forum.ipxe.org/showthread.php?tid=7807. Since if I run Windows directly from iscsi, the Windows logo hangs for about 1 minute. I tried the suggested solutions in this post, didn't help
> 
> @BorisDibrov Can you be specific about which suggested solutions you have tried?

https://forum.ipxe.org/showthread.php?tid=7807&pid=11871#pid11871

**Reply by mcb30** [reply]
> > @BorisDibrov Can you be specific about which suggested solutions you have tried?
> 
> https://forum.ipxe.org/showthread.php?tid=7807&pid=11871#pid11871

That post refers to at least two different tests.  Which **exact** modification did you make to the source code, and what was the result that you saw?

**Reply by boris-dibrov** [reply]
> > > @BorisDibrov Can you be specific about which suggested solutions you have tried?
> > 
> > 
> > https://forum.ipxe.org/showthread.php?tid=7807&pid=11871#pid11871
> 
> That post refers to at least two different tests. Which **exact** modification did you make to the source code, and what was the result that you saw?

```
--- a/src/drivers/net/realtek.c
+++ b/src/drivers/net/realtek.c
@@ -988,8 +988,8 @@ static void realtek_poll ( struct net_device *netdev ) {

        /* Check for and acknowledge interrupts */
        isr = readw ( rtl->regs + RTL_ISR );
-       if ( ! isr )
-               return;
+       //      if ( ! isr )
+       //              return;
        writew ( isr, rtl->regs + RTL_ISR );

        /* Poll for TX completions, if applicable */
@@ -997,7 +997,7 @@ static void realtek_poll ( struct net_device *netdev ) {
                realtek_poll_tx ( netdev );

        /* Poll for RX completionsm, if applicable */
-       if ( isr & ( RTL_IRQ_RER | RTL_IRQ_ROK ) )
+       //      if ( isr & ( RTL_IRQ_RER | RTL_IRQ_ROK ) )
                realtek_poll_rx ( netdev );

        /* Check link state, if applicable */
```

Modifications:
-  by removing "if ( ! isr ) return;" - **Nothig change**
 - by removing "if ( isr & ( RTL_IRQ_RER | RTL_IRQ_ROK ) )" - **Nothig change**
 - by removing "if ( ! isr ) return;" and "if ( isr & ( RTL_IRQ_RER | RTL_IRQ_ROK ) )" -  **Nothig change**
 ---
 Not a single change in the code led to any results, everything is the same. Also you said in this post ```make bin-x86_64-pcbios/realtek.pxe```. I tried it, only I got an error when build `error: code model kernel does not support PIC mode`. I'm add in **Makefile.housekeeping**
```CFLAGS          += $(WORKAROUND_CFLAGS) $(EXTRA_CFLAGS) -fno-pie```

All these actions gave nothing, the same result. Transfer speed are high on some and low on others
